cairo/crates/token/src/extensions/hyp_erc20_dex_collateral.cairo

```diff
@@ -0,0 +1,239 @@
+use starknet::ContractAddress;
+
+#[starknet::interface]
+pub trait IHypErc20DexCollateral<TContractState> {
+    fn get_dex(self: @TContractState) -> starknet::ContractAddress;
+    fn get_deposit_token(self: @TContractState) -> starknet::ContractAddress;
+    fn balance_on_behalf_of(self: @TContractState, account: ContractAddress) -> u256;
+}
+
+#[starknet::contract]
+mod HypErc20DexCollateral {
+    use alexandria_bytes::Bytes;
+    use contracts::client::gas_router_component::GasRouterComponent;
+    use contracts::client::mailboxclient_component::MailboxclientComponent;
+    use contracts::client::router_component::RouterComponent;
+    use contracts::utils::utils::U256TryIntoContractAddress;
+    use core::array::ArrayTrait;
+    use openzeppelin::access::ownable::OwnableComponent;
+    use openzeppelin::token::erc20::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait};
+    use openzeppelin::upgrades::interface::IUpgradeable;
+    use openzeppelin::upgrades::upgradeable::UpgradeableComponent;
+    use starknet::{ContractAddress, syscalls::call_contract_syscall};
+    use token::components::{
+        hyp_erc20_collateral_component::HypErc20CollateralComponent,
+        token_router::{
+            TokenRouterComponent, TokenRouterComponent::MessageRecipientInternalHookImpl,
+            TokenRouterComponent::{TokenRouterHooksTrait}, TokenRouterTransferRemoteHookDefaultImpl,
+        },
+    };
+
+    // selector for "deposit_on_behalf_of" function in the DEX contract
+    const DEX_DEPOSIT_ON_BEHALF_OF_SELECTOR: felt252 =
+        152884417735717128974538630286950396387019428546378603946454937413393931990;
```

Comment:
can you add a link to where this is from?

Reply:
This is starknet's variant of keccak

> Starknet’s version of the Keccak hash function, commonly denoted by sn_keccak
It is defined as the first 250 bits of Ethereum’s keccak256

 https://docs.starknet.io/architecture-and-concepts/cryptography/